### PR TITLE
SPARK-2149. [MLLIB] Univariate kernel density estimation

### PR DESCRIPTION
**Pull request by sryza**

### CODE REVIEW

**Comment by srowen** [comment]
@sryza are you interested in getting this in? @mengxr what's your opinion of adding this? My last review comment is that we should just use `Math` instead of `FastMath` as the latter isn't used elsewhere in Spark. Unless there's a clear performance reason for doing so. In which case, hey, let's use the fast version everywhere.

**Comment by mengxr** [comment]
Does `FastMath` give significant performance improvement over `math`? I think this needs some performance testing, given that there are other overheads involved in the computation. If we don't see significant gain, maybe it is not worthing using it. People might have different versions of commons-math3 on the classpath, so we should try to use a minimal subset of its functions. From its doc (http://commons.apache.org/proper/commons-math/apidocs/org/apache/commons/math3/util/FastMath.html), many functions are marked as "since 3.4". Another issue is the optimization used in FastMath. For example, this is the `log` implementation:

https://github.com/apache/commons-math/blob/master/src/main/java/org/apache/commons/math3/util/FastMath.java#L1141

It might be faster than JVM's implementation. But if anything (accuracy/performance) goes wrong there, it will be extremely hard for us the trace the problem.

About the API, is it okay to put `kernelDensity` as a method under `Statistics` and hide the implementation?

**Comment by sryza** [comment]
Sorry for the delay on this @mengxr @srowen.  Updated patch to take out FastMath and expose the method in `Statistics`.

**Comment by srowen** [comment]
Looking OK to me. I'll wait a beat for @mengxr to add any final comments.

File: mllib/src/main/scala/org/apache/spark/mllib/stat/KernelDensity.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+   */
+  def estimate(samples: RDD[Double], standardDeviation: Double,
+      evaluationPoints: Array[Double]): Array[Double] = {
+    if (standardDeviation <= 0.0) {
+      throw new IllegalArgumentException("Standard deviation must be positive")
+    }
+
+    // This gets used in each Gaussian PDF computation, so compute it up front
+    val logStandardDeviationPlusHalfLog2Pi =
+      Math.log(standardDeviation) + 0.5 * Math.log(2 * Math.PI)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/stat/KernelDensity.scala]: `Math` is being deprecated. Please replace it with `math` instead.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/stat/KernelDensity.scala]: Ah, yeah I thought of that before I merged, but saw a load of usages of `Math` in the code. Shall I make a PR to change all of them in one go?

**Pull request closed without merging**